packages/vite/src/node/cleanString.ts

```diff
@@ -0,0 +1,31 @@
+// bank on the non-overlapping nature of regex matches and combine all filters into one giant regex
+// /`([^`\$\{\}]|\$\{(`|\g<1>)*\})*`/g can match nested string template
+// but js not support match expression(\g<0>). so clean string template(`...`) in other ways.
+const cleanerRE = /"[^"]*"|'[^']*'|\/\*(.|[\r\n])*?\*\/|\/\/.*/g
+
+const blankReplacer = (s: string) => ' '.repeat(s.length)
+const stringBlankReplacer = (s: string) =>
+  `${s[0]}${'\0'.repeat(s.length - 2)}${s[0]}`
+
+export function emptyCommentsString(raw: string): string {
+  return raw.replace(cleanerRE, (s: string) =>
+    s[0] === '/' ? blankReplacer(s) : s
+  )
+}
+
+export function emptyString(raw: string): string {
+  const res = raw.replace(cleanerRE, (s: string) =>
+    s[0] === '/' ? blankReplacer(s) : stringBlankReplacer(s)
+  )
+  return res
+}
+
+export function findEmptyStringRawIndex(
+  clean: string,
+  emptyFlag: string,
+  start: number
+): [number, number] {
+  const flagIndex = clean.indexOf(emptyFlag, start)
+  const flagEndIndex = flagIndex + emptyFlag.length
+  return [flagIndex, flagEndIndex]
+}
```

Comment:
I think we can generalise this as
```js
export function findFlagIndex(
   str: string,
   flag: string,
   start: number
 )
```

But I also feel like it could be better to not abstract this function instead, since it's general string manipulation, and it's easier to understand the code flow if these lines are inlined in the consumer code.

Happy to hear if others have thoughts on this.

Reply:
or we can add a funtion return an object in `emptyString` to catch the original string that was replaced by `emptyString`.


like this

```ts
interface EmptyStringWithRawResult {
  clean: string
  rawStrings: Record<number, string>
}

export function emptyStringWithRaw (raw: string): EmptyStringWithRawResult {
  return {
     clean: '"\0\0\0\0"',
     rawStrings: {
        1: 'aaaa'
     }
  }
}
```

Reply:
I think my concern is that the function is a bit simple that it makes sense to write the code directly instead, at 

https://github.com/vitejs/vite/blob/a74bf6c128ff8a6101c4c788cdbd3bf73d06a263/packages/vite/src/node/plugins/assetImportMetaUrl.ts#L36-L40

and 

https://github.com/vitejs/vite/blob/a74bf6c128ff8a6101c4c788cdbd3bf73d06a263/packages/vite/src/node/plugins/workerImportMetaUrl.ts#L122-L126

So they would be:

```js
const urlStart = cleanString.indexOf(emptyUrl, index)
const urlEnd = urlStart + emptyUrl.length
```

without using the function.

This is probably my preference after reading much of Vite's code, but it's a nitpick of mine that you can ignore if you prefer too 😬

Reply:
emm....  Now `cleanString` handles the cleaning of strings and comments. It seems that these methods can also be used without encapsulation. Because the result obtained through `emptyString` must be found in the original string. It's better not to encapsulate now 😊